### ISSUE TITLE
Figure a way to make the field-as-block feature more discoverable.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#2306]
None of the layouts that we ship with core has `node/%` as path, thus people trying to add a field as a block won't be able to figure it out unless they have good knowledge of the internals of layout paths/contexts. This is something that has made me scratch my head more than once and it honestly is a shame to have in-built such an amazing feature, but not expose it prominently or at least in a way that's easy to figure out.

Possible solutions:

1. We could ship core with another layout that has its path set to `node/%` and includes a field block. A good candidate would be a "Post" layout that has the visibility set to only post nodes and the main content is a display without the image field so we can have that as a field block placed in a sidebar region.

2. We could show all blocks in the block selection dialog and have the ones that are not available be in a disabled state (grayed out) with a note in the form of "This block is available only in layouts that have the *Node* context. For example their path is set to `node/%`". That should be enough to give people a hint as to why the blocks they are looking for are not available and also as to what they need to do.

3. ???

**Comment by docwilmot** [comment]
> We could show all blocks in the block selection dialog and have the ones that are not available be in a disabled state (grayed out) with a note 

I like that one best. But perhaps we should just do both. Layouts we say are one of our best features, and we should ship with it on full(er) display. Also a layout for `user/%` with user image block in the sidebar, and/or for `/contact`.

**Comment by klonos** [comment]
> Also a layout for user/% with user image block in the sidebar, and/or for /contact.

Yes! 👍 for more default layouts with core that users can use as examples or easily tweak (instead of having to create from scratch). I'll file a separate issue for that.

**Comment by olafgrabienski** [comment]
> None of the layouts that we ship with core has node/% as path, thus people trying to add a field as a block won't be able to figure it out unless they have good knowledge of the internals of layout paths/contexts.

Additionally to the suggestions, we could improve the help text at the page where a layout path is defined. It's kind of cryptic and made me scratch my head as well. (Better a separate issue for that?)

**Comment by docwilmot** [comment]
Worked with blocks again last evening, fixing https://github.com/backdrop/backdrop-issues/issues/1704, and I'd say this issue here is a good one. Its a great, but really hidden feature. Can we have a chime in on the options?
1. ship core with another layout that has its path set to node/% and includes a field block.
2. show all blocks in the block selection dialog and have the ones that are not available be in a disabled state (grayed out)
@everybody

**Comment by docwilmot** [comment]
Pushed PR trying out 2nd option.

**Comment by docwilmot** [comment]
![show all blocks](https://cloud.githubusercontent.com/assets/1042283/20006559/0a5b7610-a266-11e6-8dff-154b0efa2387.PNG)

**Comment by klonos** [comment]
Thanx @docwilmot. I took a look at the PR sandbox and I think that this is an improvement that should get in, BUT...

1. This solution will only help people like you and me that are aware of this issue/limitation and also annoyed by it so much so to remember what they need to do next. It tells the user what they cannot do, but doesn't give them options or guidance as to what they need to do next. AFAIK, all other places in the admin UI that have disabled things give you the option to enable them.

  A way to work around this would be to create the respective documentation on b.org and link to it from the help text/description of these "disabled" fields.

  Nevertheless, this is better UX than having people wondering where the hell the feature we advertise is, but still, not a good UX as we are leading the user to a dead-end. Ideally, there would be some "add required contexts" button/link there that would "magically" do that for the user.


2. I would still like us to add a new layout in core with `node/%` as context, because consider the current situation:

  - we advertise the fields-as-blocks feature
  - we ship core with the home, admin and default layouts,
  - none of these layouts have `node/%` as context, so all nodes fall back to using the default layout
  - so, if one wants to add a field as a block, they have no way to do that out-of-the-box.
  - If they are knowledgeable, they'll figure it out, but still we have them do extra work so they'll think that Backdrop is "hard".
  - If they are new-comers or just less savvy, they are likely to give up trying. If that feature was one of the main reasons why they were sold on Backdrop, they are more likely to give up Backdrop altogether.

  By providing a node layout out of the box, we do some of the hard/tedious work for them -> UX++
  By having a field already placed as a block in that layout (articles with their image field on a sidebar make an excellent option), we give them an example that they can tweak or copy.

3. I would still like us to revert the process of *define context* -> *have features available* if we can figure out an intuitive way. People do not understand contexts and in my head this is like for example to first have them manually join database tables before using relationships in views.

**Comment by klonos** [comment]
I might give the 1st option a go if nobody beats me to it. That way, if we can get both PRs committed, I believe we'll have improved the situation a great deal 😉

**Comment by klonos** [comment]
...although, I find this a very big usability issue and if we nail it, then I think it shouldn't wait till 1.6.0.

**Comment by docwilmot** [comment]
> A way to work around this would be to create the respective documentation on b.org and link to it from the help text/description of these "disabled" fields.

I think this is the primary option here. 

Personally I dont think the situation is as cryptic as you describe here. Its not intuitive, so requires access to an explanation, but the explanation is simple. "You cant put a block that needs a node on a page that doesnt have a node". That is the essence of the context thing. Its quite sensible IMHO. 

> Ideally, there would be some "add required contexts" button/link there that would "magically" do that for the user.

Of course once we've explained in the docs and linked them, the user will recognize this is impossible. No button/link can add a context to a page. You have to create the page with a context. It has to already have a `node` or `user` etc on it.

> I would still like us to add a new layout in core with node/% as context

Having said all that though, I still agree we should have many more out-the-box layouts predefined, and I agree it would help to understand contexts a bit.

**Comment by quicksketch** [comment]
Thanks @docwilmot for the PR and @klonos for the feedback. The code in the PR looks great and in itself is a nice cleanup of our existing code to make it a bit more readable.

I am concerned though that showing blocks even when they're not possible to be added could result in additional confusion. It's possible to have dozens or hundreds of available blocks that aren't applicable in certain situations, and while node fields are the most obvious, you'd also end up showing blocks for any block that requires context. A strange out-of-box situations where this occurs with with the "Main page content" block, which requires an internally-defined context of overriding an existing path. So if you create a custom path for the layout, you get this confusing message:

![screenshot from 2016-11-15 15-45-17](https://cloud.githubusercontent.com/assets/100206/20329075/050e0fc2-ab4b-11e6-8729-4ff2c8843ee4.png)

Now it appears as though the block may be required, but it's not possible to add it. Fixing the messaging (or just hiding this particular block) would be necessary at the very least before accepting this PR.

**Comment by docwilmot** [comment]
How about a switch somewhere to "hide blocks which cannot be placed"? Experienced users can then turn that off. 

What would be nice as well is if a few more persons chimed in. I'm actually not 100% certain either.

**Comment by jenlampton** [comment]
Wow, great conversation going on here.  Thanks @docwilmot and @klonos for your continuted commitment to better ux!

I love the idea of providing a `user/%` layout as well as a `node/%' layout, but those should be separate issues (if we do not have them already?)

At first I was hesitant to show all the available blocks in the block list for the same reason as @quicksketch, there will be so many useless blocks in that list!  We'll be hiding what is possible by cluttering the list with things that are not possible.

But the more I think about it, the more I like the idea - especially if we can add a checkbox for the pros to hide what's not possible. I agree with @klonos that the language should be changed from saying what's not possible to saying what is possible, maybe something like `Can only be placed on a layout with a node context available, i.e. one for the path node/%`. 

But it seems like we're really close! Leaving the tag for 1.6 since this one seems attainable :)

**Comment by docwilmot** [comment]
> maybe something like Can only be placed on a layout with a node context available, i.e. one for the path node/%

The syntax I chose was because blocks can have more than one required context, and because not all context are path related, so "one for the path node/%" would not be always appropriate. Compromise: `'Can only be placed on a layout with the following contexts available: !contexts'` (where `!contexts` is a comma-separated list). That OK?

**Comment by docwilmot** [comment]
An issue to consider: how to exclude Multifield's subfields from generating Field Blocks. They should only be displayed as a group, not individually. Example a multifield called "Group Field One" has two attached fields "FieldA" and "FieldB". Only "Group Field One" should show in the block list. But all three show up with this PR.

Cant figure.

**Comment by jenlampton** [comment]
> The syntax I chose was because blocks can have more than one required context, and because not all context are path related, so "one for the path node/%" would not be always appropriate. Compromise: 'Can only be placed on a layout with the following contexts available: !contexts' (where !contexts is a comma-separated list). That OK?

The reason I objected to the suggested language (though technically correct) is because most people won't have any idea how to make `the following contexts available`.  I think it's very important to provide an example. We can change the 'i.e.' into a 'for example' if that makes you more comfortable, but we do need to tell people *how* to do it in this message.

People who are using more than one context will know how to make contexts available, this message is for people who don't.

**Comment by docwilmot** [comment]
Required contexts are defined in `hook_block_info()`. If I define a block which requires a `chef` object and a `blender` object, it would currently show up in the block list  of layouts with path (for example) `kitchen/%chef/mealtime/hungry/%blender` and `jen/%chef/likes/%blender`.

Perhaps an example could be "`Can only be placed on a layout with !list_of_contexts contexts available, for example one for the path blender/%/chef/%`; although this would be also confusing (since this is a fake path) but at least may illustrate the point.

Furthermore not all contexts have anything to do with a path (core provides an `overrides_path` context for example). So if a layout provides a path-less context we shouldnt give an example that looks like a path. `example one for the path overrides_path/%` is totally bogus.

So I think my syntax although unhelpful (I know) is likely to be least confusing. Perhaps a link to the docs definition of context? With examples?

**Comment by jenlampton** [comment]
You're still worrying about the 1% use case :)

**Comment by docwilmot** [comment]
I suspect you're still misunderstanding.

**Comment by jenlampton** [comment]
It's far, far more important that this text be clear, useful, and informative for the 99% use case. That's limited the contexts provided by core (all of which have designated paths). We don't need to make the text apply to all 100% of situations because people in that last 1% aren't going to be reading it anyway.

**Comment by jenlampton** [comment]
Here's an idea: If you are worried about the text applying to all blocks with context, why don't we just special-case the contexts provided by core with a switch statement, and only add the help text for each of them?

**Comment by docwilmot** [comment]
I was never disagreeing with you. The "misunderstanding" comment is simply because my last couple posts have been specifically about *multiple* contexts and *non-path* contexts. If there are multiple, we'll need to *list* them. I can handle the simple `node` ones easily enough. But if the 1% install a block that requires two contexts we still need to give them feedback that their block will not be placeable if the *both* contexts are not in the path. Hence: `Can only be placed on a layout with !list_of_contexts contexts available, for example one for the path blender/%/chef/%`. A list of required contexts.

**Comment by jenlampton** [comment]
> my last couple posts have been specifically about multiple contexts and non-path contexts.

This is still a 1% use case, but was included in my examples above for singular vs plural, as well as my last recommendation.  Any thoughts on those?

**Comment by quicksketch** [comment]
>  But if the 1% install a block that requires two contexts we still need to give them feedback that their block will not be placeable if the both contexts are not in the path.

Though perhaps this won't be the case for much longer, thanks to your work on https://github.com/backdrop/backdrop-issues/issues/2134. This would make providing example (not yet existing) paths a less-advisable idea, because even if a block required both a "node" and "user" context, the more common situation would be a regular `node/%` path with a relationship to the "user" context, rather than a path that directly contained both contexts.

**Comment by quicksketch** [comment]
In any case, let's postpone this to the next release. There are still challenges here to be worked out from a UX perspective beyond just the help text.

**Comment by jenlampton** [comment]
This PR was so close I wanted to see if we could push it across the finish line in 1.7. I've rebased and made some changes to the help text for pointing out what is required, and how to get it:
<img width="594" alt="screen shot 2017-04-30 at 3 47 56 pm" src="https://cloud.githubusercontent.com/assets/397895/25568661/862b5454-2dbc-11e7-81b5-f5d3e28a6ef2.png">

**Comment by quicksketch** [comment]
I posted some feedback to the PR: https://github.com/backdrop/backdrop/pull/1854#pullrequestreview-35653819

A very odd side-effect is that the quick-filter at the top of the listing doesn't include fields that have unmet context requirements. For example I can search for "Image", and I expect to see the `Field: Image (field_image)` result. But it is excluded. Then if I scroll down the list, I can find it there after all (but can't click on it). I find this to be quite confusing. If we're going to show everything in the list, the quick filter should use what's in the list.

Overall though, I'm still not certain about this. This will result in the flooding of dozens to hundreds of blocks on a site of reasonable size. I do think it is better for the out-of-box and new-user experience though.

Is adding an option a suitable compromise? We can stash it away under `admin/structure/layouts/settings`. We could also put a checkbox directly within the dialog (e.g. Include blocks missing contexts) but I would prefer not to clutter the everyday-UI with such an option.

**Comment by docwilmot** [comment]
> Is adding an option a suitable compromise?

I think that should be a must instead of an option.

**Comment by jenlampton** [comment]
> Is adding an option a suitable compromise? 

I agree @docwilmot, let's add the checkbox as an option.

> We can stash it away under admin/structure/layouts/settings. 

I like this location too 👍 

I'm adding the needs-feedback for some more opinions, but I think we're on the right track. (Also, descoping for 1.7 since we are only 4 days from release)

**Comment by olafgrabienski** [comment]
> None of the layouts that we ship with core has node/% as path, thus people trying to add a field as a block won't be able to figure it out unless they have good knowledge of the internals of layout paths/contexts. (...)

> I love the idea of providing a user/% layout as well as a `node/%' layout, but those should be separate issues (if we do not have them already?)

Working more with the Backdrop layout system lately, I realized that I need a `node/%' layout on nearly every of my projects. Sometimes I need it for a field block, but there also other things you don't get without the context, e.g. Node ID related conditions.

@klonos, you suggested to ship core with a `node/%` layout, but then this issue went in another direction. Are you planning to open a separate issue for it? Would appreciate it! I could do it as well but I'm still new to layouts (and never used Panels in Drupal), so the description would be quite basic.

**Comment by jenlampton** [comment]
We're only 1 day from 1.8 code freeze, and this hasn't seen a lot of activity, however, I think a solution for this problem might be included with our layout-relationships issue, since that will be adding some UX changes for selecting blocks. I'm going to bump the milestone on this issue, but hopefully layout relationships will help bring us a solution sooner.

**Comment by klonos** [comment]
I would like to revive this as a serious UX bug.

I have tested the current PR and it seems to work as expected. Minor suggestion is to try to match those requirement divs as close as possible to the respective "disabled" spans used in the module listing page (`.admin-disabled`):

- remove the bold font from the "Requires"
- change the font color of the "Requires" to `#800800`
- change the font color of the field label to `#999`
- change the font color on the rest of the text to `#666`
- change the font size to `0.9em`

<img width="619" alt="screen shot 2019-02-08 at 1 31 12 pm" src="https://user-images.githubusercontent.com/2423362/52455878-501aa600-2ba6-11e9-9752-6f2a75c925c7.png">

> We can stash it away under admin/structure/layouts/settings.

I often find myself scratching my head when I cannot find those fields, so hiding that setting out of context would not improve things unless that setting (to show all available blocks) is enabled by default OOTB.

**Comment by klonos** [comment]
@jenlampton care to refresh the PR sandbox? Thanks

**Comment by klonos** [comment]
>@klonos, you suggested to ship core with a `node/%` layout, but then this issue went in another direction. Are you planning to open a separate issue for it? Would appreciate it!

Sure @olafgrabienski, I have created 2 issues; one for shipping with a layout with `node/%` context, and one with a `user/%` context.

**Comment by stpaultim** [comment]
I think that the current "Custom Contexts for Layouts" issue will help make field blocks more discoverable (at the margins). 
https://github.com/backdrop/backdrop-issues/issues/3750#issuecomment-660579733

**Comment by stpaultim** [comment]
There was talk about having a option to hide blocks that are not available on the layout settings page: admin/structure/layouts/settings

I'm not seeing that option in current PR. I really like the way that this makes field blocks much easier to find, but I'm also concerned about cluttering the UI with blocks. I'm imagining a site that has 20-30 different custom fields and the implications for having one block for every custom field in the "add blocks" list every time, even when they are not useful.

**Comment by jenlampton** [comment]
> I think that the current "Custom Contexts for Layouts" issue will help make field blocks more discoverable

@stpaultim the UI changes recommended in #3750 should happen in a separate issue -- do you want to use this one or do it in a new issue so that they remain unblocked by the recommendations here (and vice versa)?

**Comment by stpaultim** [comment]
Ignore my comments about #3750 - apparently, I'm very confused about that issue.
